### PR TITLE
Add sysctl allowList

### PR DESCRIPTION
**Pull request by mmirecki**
The tuning-cni allows to set any sysctl on the created container. The admin might however want to restrict the user to a predefined whitelist of sysctls.
This PR is a POC with a proposal on how this could be implemented.

Tuning-cni would try to read a whitelist file from <cni bin dir>/tuningwhitelist/sysctlwhitelist (NOTE: temporary location only), and check the incoming sysctls agains it. Cni would return error if an unathorized sysctl is detected. 
The tuning-cni would behaves as before is the file is missing.

The POC is meant to only show the concept, the actual implementation will probably differ.

### CODE REVIEW

**Comment by s1061123** [comment]
Hi @mmirecki  It seems to be good feature for tuning/sysctl, especially Kubernetes with multus.

- Regarding whitelist/blacklist config file, current PR assumes that the same path of `tuning` however, I suppose user want to put it in different path from executable binary. 
- The path of whitelist/blacklist could be configurable, I suppose. Of course, we could have default path for that, e.g. /etc/cni/tuning/whitelist.conf or some.
- How about to use regexp for the sysctl list, instead of simple list of sysctl?

**Comment by mmirecki** [comment]
> Hi @mmirecki It seems to be good feature for tuning/sysctl, especially Kubernetes with multus.
> 
> * Regarding whitelist/blacklist config file, current PR assumes that the same path of `tuning` however, I suppose user want to put it in different path from executable binary.
> * The path of whitelist/blacklist could be configurable, I suppose. Of course, we could have default path for that, e.g. /etc/cni/tuning/whitelist.conf or some.
I've set this as the default one. I'll add an override when suggestions appear.

> * How about to use regexp for the sysctl list, instead of simple list of sysctl?
Done

**Review by s1061123 — COMMENTED** [review]
As far as I looked, the code seems good to me. I recommend unit-test for the code, in https://github.com/containernetworking/plugins/blob/master/plugins/meta/tuning/tuning_test.go

File: plugins/meta/tuning/tuning.go [hunk 1]
```diff
@@ -477,3 +484,56 @@ func cmdCheck(args *skel.CmdArgs) error {
...
+			return false, err
+		}
+		if match {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func readWhitelist() (bool,[]string, error) {
```
[the remark this answers is not in the document]
Comment (mmirecki) [reply plugins/meta/tuning/tuning.go:521]: Done

File: plugins/meta/tuning/tuning.go [hunk 2]
```diff
@@ -477,3 +484,56 @@ func cmdCheck(args *skel.CmdArgs) error {
...
+			return err
+		}
+		if !match {
+			return errors.New(fmt.Sprintf("Sysctl %s is not allowed. Only the following sysctls are allowed: %+v", sysctl, whiteList))
+		}
+	}
+	return nil
+}
+
+func contains(sysctl string, whiteList []string) (bool, error) {
```
[the remark this answers is not in the document]
Comment (mmirecki) [reply plugins/meta/tuning/tuning.go:508]: Done

File: plugins/meta/tuning/tuning.go [hunk 3]
```diff
@@ -477,3 +484,56 @@ func cmdCheck(args *skel.CmdArgs) error {
 
 	return nil
 }
+
+func validateConf(tuningConf *TuningConf) error {
```
[the remark this answers is not in the document]
Comment (mmirecki) [reply plugins/meta/tuning/tuning.go:488]: Done

**Comment by mmirecki** [comment]
> As far as I looked, the code seems good to me. I recommend unit-test for the code, in https://github.com/containernetworking/plugins/blob/master/plugins/meta/tuning/tuning_test.go

@s1061123 , added tests, replied to other comments
Thanks

**Comment by dcbw** [comment]
@mmirecki I agree with @cgoncalves ; would you mind whitelist -> allowlist?

**Comment by mmirecki** [comment]
> @mmirecki I agree with @cgoncalves ; would you mind whitelist -> allowlist?

-> allowlist

File: plugins/meta/tuning/tuning.go [hunk 4]
```diff
@@ -477,3 +485,60 @@ func cmdCheck(args *skel.CmdArgs) error {
...
+			return errors.New(fmt.Sprintf("Sysctl %s is not allowed. Only the following sysctls are allowed: %+v", sysctl, allowlist))
+		}
+	}
+	return nil
+}
+
+// Validate the whiteList contains the given sysctl
+func contains(sysctl string, whiteList []string) (bool, error) {
+	for _, whiteListElements := range whiteList {
+		match, err := regexp.MatchString(whiteListElements, sysctl)
```
Comment (s1061123) [inline plugins/meta/tuning/tuning.go:514]: whitelist -> allowlist

File: plugins/meta/tuning/tuning.go [hunk 5]
```diff
@@ -477,3 +485,60 @@ func cmdCheck(args *skel.CmdArgs) error {
...
+			return false, err
+		}
+		if match {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// Read the systctl allowlist from file. Return info if the file is present and the read whiteList if it is
```
Comment (s1061123) [inline plugins/meta/tuning/tuning.go:525]: whitelist -> allowlist

File: plugins/meta/tuning/tuning.go [hunk 6]
```diff
@@ -477,3 +485,60 @@ func cmdCheck(args *skel.CmdArgs) error {
...
+
+	lines := strings.Split(string(dat), "\n")
+	whiteList := []string{}
+	for _, line := range lines {
+		line = strings.TrimSpace(line)
+		if len(line) > 0 {
+			whiteList = append(whiteList, line)
+		}
+	}
+	return true, whiteList, nil
```
Comment (s1061123) [inline plugins/meta/tuning/tuning.go:543]: whitelist -> allowlist

**Review by s1061123 — APPROVED** [review]
It looks good to me. Thank you for the changes!

**Comment by squeed** [comment]
Looks good to me. When you have a moment, can you file a PR in https://github.com/containernetworking/cni.dev/ documenting this? Thanks

**Comment by mmirecki** [comment]
> Looks good to me. When you have a moment, can you file a PR in https://github.com/containernetworking/cni.dev/ documenting this? Thanks

@squeed Here's the doc pr: https://github.com/containernetworking/cni.dev/pull/99

**Pull request merged**